### DISCUSSION TITLE
`with_loader_criteria` - closure variable value seems to be cached

### DISCUSSION THREAD

**Discussion by ldirer** [sqlalchemy/sqlalchemy#8398] [Usage Questions]
Hello,

I have been trying to use `do_orm_execute` to automatically filter some tables by a 'current account' (to do something called 'tenant isolation').   
I ran into an issue where the value of the 'current account' filter in the emitted SQL wasn't always what I expected, as if it was cached from previous queries.  

For context I put what I'm attempting to do in an example [here](https://github.com/sqlalchemy/sqlalchemy/commit/c2da07bfb446a4d4c754b8aec19834f5f0f7f70d).   
It's very similar to the `filter_public.py` example, except the filter is on an `account_id` property and the value comes from a variable in the closure.  

I tried to narrow down the issue and was able to reproduce it in a test.  
The test uses `with_loader_criteria` with a lambda that has a closure variable (adapted from `test/orm/test_relationship_criteria.py::LoaderCriteriaTest::test_select_selectinload_mapper_mapper_criteria`):  
[https://github.com/sqlalchemy/sqlalchemy/commit/b300d188e0d4f1715921f0a157e128e0b0e06c32](https://github.com/sqlalchemy/sqlalchemy/commit/b300d188e0d4f1715921f0a157e128e0b0e06c32)

Observations:

* If I run the example with `query_cache_size=0` on the engine, it works (instead of failing on an assert).
* Similarly adding in the test adding a `s.bind.clear_compiled_cache()` line between the two executions makes the test pass.  

1. Can this be considered a bug? I expected the test to pass 'out of the box', but I could have missed something or just have crazy expectations :upside_down_face: .
2. I tried to debug this but got a bit stuck. I looked into cache keys but these do not include the values of bind parameters (so they are not in cause).  
I'd like to understand where in the code the bind parameters are passed when the query is cached but didn't find it.  
What would be the relevant tests/code to look at?

Thanks!

Versions: 1.4.35 and recent `main`, Python 3.9.

**Comment by zzzeek** [comment]
probably a bug (sorry for the big language about "dont post bugs", the people who always post bugs are the ones that should be asking questions, and the ones that ask questions are the ones that should be posting bugs :)  )  but before trying it what has me curious is that the test you have there does not seem different from an existing test we have right here : https://github.com/sqlalchemy/sqlalchemy/blob/a134ec1760df6295d537ff63df7aee83d957bf6a/test/orm/test_relationship_criteria.py#L849-L869  that's the same test?   I will try to look now to see what's different in your case...

**Reply by zzzeek** [reply]
its the loader strategies, that's where the cached version is carrying over towards.   will have to see if there's a way to affect that

**Reply by ldirer** [reply]
Thank you for your answer.  
I hadn't noticed this test, it does look very similar!  

Something that I didn't mention (I wasn't rigorous enough in my notes to be 100% sure of it) is that I think the issue occurs only when there's a `selectinload` (as opposed to a single `select`).

**Reply by zzzeek** [reply]
yes selectinload or lazyload .   this led to a great refinement of how this particular area works, new reviews for 1.4/2.0 added at #8399.
